deepinv/utils/plotting.py

```diff
@@ -115,27 +115,79 @@ def prepare_images(x=None, y=None, x_net=None, x_nl=None, rescale_mode="min_max"
     :return: the preprocessed image.
     """
-    with torch.no_grad():
-        if im.shape[1] == 2:  # for complex images
-            pimg = im.pow(2).sum(dim=1, keepdim=True).sqrt().type(torch.float32)
-        elif im.shape[1] > 3:
-            pimg = im.type(torch.float32)
-        else:
-            if torch.is_complex(im):
-                pimg = im.abs().type(torch.float32)
-            else:
-                pimg = im.type(torch.float32)
+    if not rescale_only:
+        # Apply the modulus function on each entry if the input is complex-valued
+        # or if it has 2 channels which are assume to represent the real and
+        # imaginary parts of a complex-valued image.
+        # NOTE: Expect a batched input
+        channels_dim = 1
+        n_channels = im.shape[channels_dim]
+        is_complex = torch.is_complex(im) or n_channels == 2
+        if is_complex:
+            # NOTE: This implementation is meant to work for complex-valued images
+            # and for real-valued images with two channels, understood as the real
+            # and imaginary part of a complex-valued image.
+            im = torch.linalg.vector_norm(im, ord=2, dim=channels_dim, keepdim=True)
+
+        # Cast image values to float32 numbers
+        im = im.type(torch.float32)
+
+    min_val = 0.0
+    max_val = 1.0
+
+    # NOTE: Rescaling a constant image between zero and one is ill-defined.
+    # Indeed, no matter the new scale, the image won't have a minimum value
+    # equal to zero and a maximum value equal to one. In this case, we
+    # choose to rescale the image so that its new value is the number in
+    # [0, 1] that is closest to its original value. This amounts to
+    # clamping it between zero and one.
+    if rescale_mode == "min_max":
+        # Compute batch-wise minimum and maximum values
+        # NOTE: Expect a batched input of shape (B, *)
+        non_batched_dims = list(range(1, im.ndim))
+        im_min = im.amin(dim=non_batched_dims, keepdim=False)
+        im_max = im.amax(dim=non_batched_dims, keepdim=False)
+
+        # Clone the image to avoid input mutations
+        im = im.clone()
+
+        # Rescale non-constant batch images between zero and one
+        indices = im_max != im_min
+        im[indices] -= im_min[indices]
+        im[indices] /= im_max[indices] - im_min[indices]
+
+        # Clamp constant batch images between zero and one
+        indices = torch.logical_not(indices)
+        im[indices] = im[indices].clamp(min=min_val, max=max_val)
+    elif rescale_mode == "clip":
+        im = im.clamp(min=min_val, max=max_val)
+    else:
+        raise ValueError("rescale_mode has to be either 'min_max' or 'clip'.")
+
+    return im
 
-        pimg = rescale_img(pimg, rescale_mode=rescale_mode)
-    return pimg
+
+@torch.no_grad
+def preprocess_img(im: torch.Tensor, rescale_mode: str = "min_max"):
+    r"""
+    Preprocesses an image tensor for plotting.
```

Comment:
Please can this have a more informative docstring? I.e. what is preprocessing if it is not just rescaling?